File: packages/wire/aead.go
package wire

import (
	"crypto/aes"
	"crypto/cipher"
	"encoding/binary"
	"fmt"
)

const u16Max = 0xffff

// FrameHeaderInput is the header fields the caller supplies; Len is set from the
// plaintext by Seal.
type FrameHeaderInput struct {
	Version  uint8
	Type     uint8
	Flags    uint8
	FileIdx  uint16
	BlockIdx uint32
	FrameOff uint32
}

// nonce builds the 12-byte GCM nonce: the direction's 4-byte salt followed by a
// big-endian u64 frame counter.
func nonce(salt []byte, counter uint64) []byte {
	out := make([]byte, aeadNonceBytes)
	copy(out, salt)
	binary.BigEndian.PutUint64(out[aeadSaltBytes:], counter)
	return out
}

func gcmFor(key []byte) (cipher.AEAD, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	return cipher.NewGCM(block)
}

// Seal encrypts plaintext into counter(8) || header(16) || ciphertext || tag(16). Counter and
// header are authenticated as GCM additional data.
func Seal(dir DirectionalKey, counter uint64, h FrameHeaderInput, plaintext []byte) ([]byte, error) {
	if len(plaintext) > u16Max {
		return nil, fmt.Errorf("frame payload %d exceeds u16 max %d", len(plaintext), u16Max)
	}
	header := encodeFrameHeader(FrameHeader{
		Version:  h.Version,
		Type:     h.Type,
		Flags:    h.Flags,
		FileIdx:  h.FileIdx,
		BlockIdx: h.BlockIdx,
		FrameOff: h.FrameOff,
		Len:      uint16(len(plaintext)),
	})
	aad := make([]byte, frameCounterBytes+len(header))
	binary.BigEndian.PutUint64(aad[:frameCounterBytes], counter)
	copy(aad[frameCounterBytes:], header)
	gcm, err := gcmFor(dir.Key)
	if err != nil {
		return nil, err
	}
	// Seal appends ciphertext||tag onto aad, giving header || ciphertext || tag.
	return gcm.Seal(aad, nonce(dir.Salt, counter), plaintext, aad), nil
}

// SealPadded pads plaintext to a power-of-two bucket size with a prefix length, sets FrameFlagPadded,
// and encrypts the padded payload under AES-GCM (V17-PR03).
func SealPadded(dir DirectionalKey, counter uint64, h FrameHeaderInput, plaintext []byte) ([]byte, error) {
	padded, err := PadPayload(plaintext)
	if err != nil {
		return nil, err
	}
	h.Flags |= FrameFlagPadded
	return Seal(dir, counter, h, padded)
}

// OpenedFrame is a decrypted frame: its parsed header and recovered plaintext.
type OpenedFrame struct {
	Counter   uint64
	Header    FrameHeader
	Plaintext []byte
}

// Open decrypts a frame produced by Seal. It verifies the GCM tag over the header AAD and
// the expected nonce, and fails if authentication fails, the frame is truncated, or the
// header Len disagrees with the ciphertext length.
func Open(dir DirectionalKey, counter uint64, frame []byte) (*OpenedFrame, error) {
	embedded, err := frameCounter(frame)
	if err != nil {
		return nil, err
	}
	if embedded != counter {
		return nil, fmt.Errorf("frame counter %d does not match expected %d", embedded, counter)
	}
	return openEmbedded(dir, embedded, frame)
}

// ErrFrameReplay identifies an already-consumed counter. Transfer engines ignore it while
// keeping all forward counters authenticated, which makes transport replacement replay-safe.
var ErrFrameReplay = Errorf(CodeProtocol, "frame counter replay")

// OpenSequenced opens a transfer frame using its authenticated wire counter. Forward gaps are
// accepted after a transport loses a suffix; counters below minimum return ErrFrameReplay.
func OpenSequenced(dir DirectionalKey, minimum uint64, frame []byte) (*OpenedFrame, error) {
	embedded, err := frameCounter(frame)
	if err != nil {
		return nil, err
	}
	if embedded < minimum {
		return nil, fmt.Errorf("%w: counter %d below %d", ErrFrameReplay, embedded, minimum)
	}
	return openEmbedded(dir, embedded, frame)
}

func frameCounter(frame []byte) (uint64, error) {
	if len(frame) < frameCounterBytes+frameHeaderBytes+aeadTagBytes {
		return 0, fmt.Errorf("frame too short: %d bytes", len(frame))
	}
	return binary.BigEndian.Uint64(frame[:frameCounterBytes]), nil
}

func openEmbedded(dir DirectionalKey, counter uint64, frame []byte) (*OpenedFrame, error) {
	aadBytes := frameCounterBytes + frameHeaderBytes
	aad := frame[:aadBytes]
	header, err := decodeFrameHeader(aad[frameCounterBytes:])
	if err != nil {
		return nil, err
	}
	body := frame[aadBytes:]
	if len(body) != int(header.Len)+aeadTagBytes {
		return nil, fmt.Errorf("frame len field %d disagrees with body %d", header.Len, len(body))
	}
	gcm, err := gcmFor(dir.Key)
	if err != nil {
		return nil, err
	}
	decrypted, err := gcm.Open(nil, nonce(dir.Salt, counter), body, aad)
	if err != nil {
		return nil, err
	}
	plaintext := decrypted
	if header.Flags&FrameFlagPadded != 0 {
		unpadded, err := UnpadPayload(decrypted)
		if err != nil {
			return nil, err
		}
		plaintext = unpadded
	}
	return &OpenedFrame{Counter: counter, Header: header, Plaintext: plaintext}, nil
}

File: packages/protocol/src/aead.ts
/**
 * AES-256-GCM frame codec. Each frame is an 8-byte monotonic counter, a fixed 16-byte header,
 * and the GCM output (ciphertext with appended 16-byte tag). Counter and header are the GCM
 * additional authenticated data, so tampering with either fails decryption.
 *
 * The 12-byte nonce is the direction's 4-byte salt followed by a big-endian u64 frame
 * counter. Counters are per-direction and supplied by the caller (the session tracks
 * them), giving nonce continuity without ever reusing a (key, nonce) pair. Uses native
 * WebCrypto; mirrored by Go `crypto/cipher` in `packages/wire`.
 */

import {
  AEAD_NONCE_BYTES,
  AEAD_TAG_BYTES,
  FRAME_COUNTER_BYTES,
  FRAME_FLAG_PADDED,
  FRAME_HEADER_BYTES,
} from './constants.js';
import { decodeFrameHeader, encodeFrameHeader, padPayload, unpadPayload } from './frame.js';
import type { DirectionalKey } from './keyschedule.js';
import type { FrameHeader } from './transfer.js';
import { aesGcmOpen, aesGcmSeal } from './webcrypto.js';

/** Header fields the caller supplies; `len` is set from the plaintext by `seal`. */
export type FrameHeaderInput = Omit<FrameHeader, 'len'>;

const U16_MAX = 0xffff;

function nonce(salt: Uint8Array, counter: number): Uint8Array {
  if (!Number.isSafeInteger(counter) || counter < 0) {
    throw new RangeError(`frame counter ${counter} is not a non-negative integer`);
  }
  const out = new Uint8Array(AEAD_NONCE_BYTES);
  out.set(salt, 0);
  new DataView(out.buffer).setBigUint64(salt.length, BigInt(counter), false);
  return out;
}

/**
 * Encrypt `plaintext` into `counter(8) || header(16) || ciphertext || tag(16)`.
 */
export async function seal(
  dir: DirectionalKey,
  counter: number,
  header: FrameHeaderInput,
  plaintext: Uint8Array,
): Promise<Uint8Array> {
  if (plaintext.length > U16_MAX) {
    throw new RangeError(`frame payload ${plaintext.length} exceeds u16 max ${U16_MAX}`);
  }
  const aad = new Uint8Array(FRAME_COUNTER_BYTES + FRAME_HEADER_BYTES);
  new DataView(aad.buffer).setBigUint64(0, BigInt(counter), false);
  aad.set(encodeFrameHeader({ ...header, len: plaintext.length }), FRAME_COUNTER_BYTES);
  const sealed = await aesGcmSeal(dir.key, nonce(dir.salt, counter), aad, plaintext);
  const out = new Uint8Array(aad.length + sealed.length);
  out.set(aad, 0);
  out.set(sealed, aad.length);
  return out;
}

/**
 * Encrypt `plaintext` with traffic padding applied (FRAME_FLAG_PADDED set) (V17-PR03).
 */
export async function sealPadded(
  dir: DirectionalKey,
  counter: number,
  header: FrameHeaderInput,
  plaintext: Uint8Array,
): Promise<Uint8Array> {
  const padded = padPayload(plaintext);
  return seal(dir, counter, { ...header, flags: header.flags | FRAME_FLAG_PADDED }, padded);
}

/** A decrypted frame: its parsed header and recovered plaintext. */
export interface OpenedFrame {
  readonly counter: number;
  readonly header: FrameHeader;
  readonly plaintext: Uint8Array;
}

/**
 * Decrypt a frame produced by `seal`. Verifies the GCM tag over the header AAD and the
 * expected nonce; throws if authentication fails, the frame is truncated, or the header
 * `len` disagrees with the ciphertext length.
 */
export async function open(
  dir: DirectionalKey,
  counter: number,
  frame: Uint8Array,
): Promise<OpenedFrame> {
  const embedded = readCounter(frame);
  if (embedded !== counter) {
    throw new Error(`frame counter ${embedded} does not match expected ${counter}`);
  }
  return openEmbedded(dir, embedded, frame);
}

/** A frame older than the next accepted counter; callers may safely ignore this replay. */
export class FrameReplayError extends Error {
  constructor(
    readonly counter: number,
    readonly minimum: number,
  ) {
    super(`frame counter ${counter} is below next accepted ${minimum}`);
    this.name = 'FrameReplayError';
  }
}

/**
 * Open a transfer frame carrying its own authenticated counter. Forward gaps are permitted so a
 * reliable transport can be replaced after losing a suffix; already-consumed counters are
 * reported as replays and nonce values are never reset or guessed.
 */
export async function openSequenced(
  dir: DirectionalKey,
  minimumCounter: number,
  frame: Uint8Array,
): Promise<OpenedFrame> {
  const embedded = readCounter(frame);
  if (embedded < minimumCounter) throw new FrameReplayError(embedded, minimumCounter);
  return openEmbedded(dir, embedded, frame);
}

function readCounter(frame: Uint8Array): number {
  if (frame.length < FRAME_COUNTER_BYTES + FRAME_HEADER_BYTES + AEAD_TAG_BYTES) {
    throw new Error(`frame too short: ${frame.length} bytes`);
  }
  const value = new DataView(frame.buffer, frame.byteOffset, FRAME_COUNTER_BYTES).getBigUint64(
    0,
    false,
  );
  if (value > BigInt(Number.MAX_SAFE_INTEGER)) throw new Error('frame counter exceeds JS range');
  return Number(value);
}

async function openEmbedded(
  dir: DirectionalKey,
  counter: number,
  frame: Uint8Array,
): Promise<OpenedFrame> {
  const aadBytes = FRAME_COUNTER_BYTES + FRAME_HEADER_BYTES;
  const aad = frame.slice(0, aadBytes);
  const header = decodeFrameHeader(aad.subarray(FRAME_COUNTER_BYTES));
  const body = frame.slice(aadBytes);
  if (body.length !== header.len + AEAD_TAG_BYTES) {
    throw new Error(`frame len field ${header.len} disagrees with body ${body.length}`);
  }
  const decrypted = await aesGcmOpen(dir.key, nonce(dir.salt, counter), aad, body);
  let plaintext = decrypted;
  if ((header.flags & FRAME_FLAG_PADDED) !== 0) {
    plaintext = unpadPayload(decrypted);
  }
  return { counter, header, plaintext };
}
